include/xsparse/levels/hashed.hpp

```diff
@@ -0,0 +1,167 @@
+#ifndef XSPARSE_LEVELS_HASHED_HPP
+#define XSPARSE_LEVELS_HASHED_HPP
+
+#include <tuple>
+#include <utility>
+#include <optional>
+#include <unordered_map>
+
+#include <xsparse/util/base_traits.hpp>
+#include <xtl/xiterator_base.hpp>
+
+namespace xsparse
+{
+    namespace levels
+    {
+        template <class LowerLevels, class IK, class PK, class CrdContainer>
+        class hashed;
+
+        template <class... LowerLevels, class IK, class PK, class CrdContainer>
+        class hashed<std::tuple<LowerLevels...>, IK, PK, CrdContainer>
+        {
+            using BaseTraits
+                = util::base_traits<hashed, std::tuple<LowerLevels...>, IK, PK, CrdContainer>;
+
+        public:
+            class iteration_helper
+            {
+                static_assert(std::is_nothrow_invocable_r_v<std::optional<typename BaseTraits::PK>,
+                                                            decltype(&BaseTraits::Level::locate),
+                                                            typename BaseTraits::Level&,
+                                                            typename BaseTraits::PKM1,
+                                                            typename BaseTraits::IK>);
+
+            private:
+                typename BaseTraits::Level const& m_level;
+                typename BaseTraits::PKM1 const m_pkm1;
+                typename BaseTraits::I const m_i;
+
+            public:
+                class iterator;
+                using value_type =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using difference_type = typename std::make_signed_t<typename BaseTraits::PK>;
+                using pointer =
+                    typename std::pair<typename BaseTraits::IK, typename BaseTraits::PK>*;
+                using reference = std::pair<typename BaseTraits::IK, typename BaseTraits::PK>;
+                using iterator_type = iterator;
+
+                class iterator : xtl::xbidirectional_iterator_base2<iteration_helper>
```

Comment:
```suggestion
                class iterator : public xtl::xbidirectional_iterator_base2<iteration_helper>
```

So we can access all methods in the base class.